src/renderer/viz/expressions/CategoryIndex.js

```diff
@@ -0,0 +1,142 @@
+import BaseExpression from './base';
+import { checkExpression, implicitCast, checkType, checkMaxArguments } from './utils';
+
+const SQRT_MAX_CATEGORIES_PER_PROPERTY = 256;
+
+/**
+* Index value for the given category in the property category list
+*
+* @param {Category} property - The property to be evaluated, must be categorical
+* @return {Category}
+*
+* @example <caption> Color by $city using the CARTOColor Prism by assigning different color in Prism to each category.</caption>
+* const s = carto.expressions;
+* const viz = new carto.Viz({
+*   color: s.ramp(s.categoryIndex(s.prop('city')), s.palettes.PRISM)
+* });
+*
+* @example <caption> Color by $city using the CARTOColor Prism by assigning different color in Prism to each category. (String)</caption>
+* const viz = new carto.Viz(`
+*   color: ramp(categoryIndex($city), PRISM)
+* `);
+*
+* @memberof carto.expressions
+* @name categoryIndex
+* @function
+* @api
+*/
+export default class CategoryIndex extends BaseExpression {
+    constructor (property) {
+        checkMaxArguments(arguments, 1, 'categoryIndex');
+
+        property = implicitCast(property);
+
+        checkExpression('categoryIndex', 'property', 0, property);
+        super({ property });
+        this._numTranslatedCategories = null;
+        this.type = 'category';
+
+        this._translatePixels = new Float32Array(SQRT_MAX_CATEGORIES_PER_PROPERTY * SQRT_MAX_CATEGORIES_PER_PROPERTY);
+        this._translateArray = [];
+        this._numTranslatedCategories = 0;
+        this._numTranslatedCategoriesGL = 0;
+    }
+
+    _bindMetadata (metadata) {
+        super._bindMetadata(metadata);
+        this._metadata = metadata;
+        const metaColumn = this._metadata.properties[this.property.name];
+        this.numCategories = metaColumn.categories.length;
+        this.numCategoriesWithoutOthers = this.numCategories;
```

Comment:
We need to set this on `_preDraw` too, since custom MVTs may not have the complete list of categories at this point